### ISSUE TITLE
twisted.python.log.showwarning doesn't take the new line argument added in Python 2.6

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3211]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3211|
|Type|defect|
|Created|2008-04-26 15:14:29Z|
|Branch|https://github.com/twisted/twisted/tree/showwarning-line-3211|

It should be done at some point, it probably speed ups some operation.

<details><summary>Searchable metadata</summary>

```
trac-id__3211 3211
type__defect defect
reporter__therve therve
priority__low low
milestone__Python_2_6 Python-2.6
branch__branches_showwarning_line_3211 branches/showwarning-line-3211
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1209222869000000 1209222869000000
changetime__1221507025000000 1221507025000000
version__None None
owner__ 
cc__exarkun cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

warnings.showwarning is a public hook.  Should we complain about this backwards incompatible change?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

That's a good point. We can always raise the issue and see what they think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I filed [a ticket](http://bugs.python.org/issue2705) in the Python bug tracker.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

The parameter is now optional. We should now manage this argument, but that's not a priority.

Old description:
```
Revision 62303 on Python trunk (http://svn.python.org/view?rev=62303&view=rev) introduced a new parameter line in showwarning to add line cache. t.p.log tries to override showwarning but misses the new parameter, so it fails.
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

These error messages also popup when you enable the Py3k warnings mode in Python 2.5 with the '-3' option.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

Err, that should be 'python 2.6' in the previous comment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [24805]) Branching to 'showwarning-line-3211'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [24808]) note about how this can't be improved right now

refs [#3433](https://github.com/twisted/twisted/issues/3433)
refs [#3211](https://github.com/twisted/twisted/issues/3211)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

Added the parameter and some tests.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

I think we should try to pass the line parameter to `_oldshowwarning` on 2.6. It probably has noticeable performance impact. Everything else looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [24857]) pass on `line` if appropriate

refs [#3211](https://github.com/twisted/twisted/issues/3211)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.

I didn't add performance tests, but it passes on the line parameter now.  The test makes up a file/line that doesn't exist so the warning text isn't even right if the line isn't passed on appropriately.  I guess this is a good thing.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

Cool, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [24858]) Merge showwarning-line-3211

Author: exarkun
Reviewer: therve
Fixes: [#3211](https://github.com/twisted/twisted/issues/3211)

Change the `warnings.showwarning` hook installed by `twisted.python.log`
so that it accepts a new, optional `line` parameter which is passed by
the Python warning system in Python 2.6.  This silences a huge number of
warnings when using Python 2.6 which are emitted when using any Twisted
code which invokes the Python warning system.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
